src/iceberg/type.cc

```diff
@@ -0,0 +1,314 @@
+std::string Float64Type::ToString() const { return "float64"; }
+bool Float64Type::Equals(const Type& other) const {
+  return other.type_id() == TypeId::kFloat64;
+}
+
+DecimalType::DecimalType(int32_t precision, int32_t scale)
+    : precision_(precision), scale_(scale) {
+  if (precision < 0 || precision > kMaxPrecision) {
+    throw std::runtime_error(
+        std::format("DecimalType: precision must be in [0, 38], was {}", precision));
+  }
+}
+
+int32_t DecimalType::precision() const { return precision_; }
+int32_t DecimalType::scale() const { return scale_; }
+TypeId DecimalType::type_id() const { return TypeId::kDecimal; }
+std::string DecimalType::ToString() const {
+  return std::format("decimal({}, {})", precision_, scale_);
+}
+bool DecimalType::Equals(const Type& other) const {
+  if (other.type_id() != TypeId::kDecimal) {
+    return false;
+  }
+  const auto& decimal = static_cast<const DecimalType&>(other);
+  return precision_ == decimal.precision_ && scale_ == decimal.scale_;
+}
+
+TypeId TimeType::type_id() const { return TypeId::kTime; }
+std::string TimeType::ToString() const { return "time"; }
+bool TimeType::Equals(const Type& other) const {
+  return other.type_id() == TypeId::kTime;
+}
+
+TypeId DateType::type_id() const { return TypeId::kDate; }
+std::string DateType::ToString() const { return "date"; }
+bool DateType::Equals(const Type& other) const {
+  return other.type_id() == TypeId::kDate;
+}
+
+bool TimestampType::is_zoned() const { return false; }
+TimeUnit TimestampType::time_unit() const { return TimeUnit::kMicrosecond; }
+TypeId TimestampType::type_id() const { return TypeId::kTimestamp; }
+std::string TimestampType::ToString() const { return "timestamp"; }
+bool TimestampType::Equals(const Type& other) const {
+  return other.type_id() == TypeId::kTimestamp;
+}
+
+bool TimestampTzType::is_zoned() const { return true; }
+TimeUnit TimestampTzType::time_unit() const { return TimeUnit::kMicrosecond; }
+TypeId TimestampTzType::type_id() const { return TypeId::kTimestampTz; }
+std::string TimestampTzType::ToString() const { return "timestamptz"; }
+bool TimestampTzType::Equals(const Type& other) const {
+  return other.type_id() == TypeId::kTimestampTz;
+}
+
+TypeId BinaryType::type_id() const { return TypeId::kBinary; }
+std::string BinaryType::ToString() const { return "binary"; }
+bool BinaryType::Equals(const Type& other) const {
+  return other.type_id() == TypeId::kBinary;
+}
+
+TypeId StringType::type_id() const { return TypeId::kString; }
+std::string StringType::ToString() const { return "string"; }
+bool StringType::Equals(const Type& other) const {
+  return other.type_id() == TypeId::kString;
+}
+
+FixedType::FixedType(int32_t length) : length_(length) {
+  if (length < 0) {
+    throw std::runtime_error(
+        std::format("FixedType: length must be >= 0, was {}", length));
+  }
+}
+
+int32_t FixedType::length() const { return length_; }
```

Comment:
an unimportant issue: should we just use uint32_t?

Reply:
Java uses int, so I'd rather be consistent

Reply:
Seems that Java not has unsigned int? But both part lgtm

Reply:
Java does not have unsigned types but you can manipulate signed types as if they were unsigned when necessary. I'll close this since there's another thread about this now...